### PR TITLE
Dolby/ac4

### PR DESCRIPTION
**Pull request by ShaoWeiguo**
Shaka Packager supports Dolby AC-4 codec (channel based, IMS and CBI), and only supports single presentation.

### CODE REVIEW

File: packager/mpd/base/media_info.proto [hunk 1]
```diff
@@ -75,6 +75,26 @@ message MediaInfo {
     // Backwards-compatible object audio carriage using Enhanced AC-3 Standard
     // C.3.2.3.
     optional uint32 ec3_joc_complexity = 3;
+
+    // AC4 Channel mask bit fields, encoded based on ETSI TS 103 190-2 V1.2.1
+    // Digital Audio Compression (AC-4) Standard; Part 2: Immersive and
+    // personalized audio E.10.14.
+    optional uint32 ac4_channel_mask = 4;
```
Comment (kqyang) [inline packager/mpd/base/media_info.proto:82]: What do you think of renaming `ec3_channel_map` to `channel_map` or `channel_mask` and use it for both EC3 and AC4?

Similarly for channel_mpeg_value.

File: packager/media/codecs/ac4_audio_util.cc [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+// 11,       Top centre
+// 12,       LFE2
+// 13,       Bottom front left/Bottom front right pair
+// 14,       Bottom front centre
+// 15,       Back centre
+// 16,       Left screen/Right screen pair
+// 17,       Left wide/Right wide pair
+// 18,       Vertical height left/Vertical height right pair
+enum kAC4AudioChannelGroupIndex {
+    kLRpair = 0x1,
```
Comment (kqyang) [inline packager/media/codecs/ac4_audio_util.cc:41]: kLRPair to be consistent with the naming below.

Comment (ShaoWeiguo) [reply packager/media/codecs/ac4_audio_util.cc:41]: You are right. Fixed.

File: packager/mpd/base/xml/xml_node.cc [hunk 3]
```diff
@@ -500,6 +501,40 @@ bool RepresentationXmlNode::AddAudioChannelInfo(const AudioInfo& audio_info) {
...
+      // Calculate AC-4 channel configuration descriptor value with MPEG scheme.
+      // Spec: ETSI TS 103 190-2 V1.2.1 Digital Audio Compression (AC-4) Standard;
+      // Part 2: Immersive and personalized audio G.3.2.
+      audio_channel_config_value = base::UintToString(ac4_channel_mpeg_value);
+      audio_channel_config_scheme = "urn:mpeg:mpegB:cicp:ChannelConfiguration";
+    }
+    bool ret = AddDescriptor("AudioChannelConfiguration",
+                             audio_channel_config_scheme,
+                             audio_channel_config_value);
+    if (ac4_ims_flag) {
```
[the remark this answers is not in the document]
Comment (ShaoWeiguo) [reply packager/mpd/base/xml/xml_node.cc:532]: Yes. ac4_cbi_flag is only used in HLS playlist. 
For dash, the manifest contains channel configuration which can represent height channels.

File: packager/mpd/base/xml/xml_node.cc [hunk 4]
```diff
@@ -500,6 +501,40 @@ bool RepresentationXmlNode::AddAudioChannelInfo(const AudioInfo& audio_info) {
...
+    if (ac4_channel_mpeg_value == NO_MAPPING) {
+      // Calculate AC-4 channel mask. Spec: ETSI TS 103 190-2 V1.2.1 Digital
+      // Audio Compression (AC-4) Standard; Part 2: Immersive and personalized
+      // audio G.3.1.
+      const uint32_t ac4_channel_mask =
+        base::HostToNet32(codec_data.ac4_channel_mask() << 8);
+      audio_channel_config_value =
+        base::HexEncode(&ac4_channel_mask, sizeof(ac4_channel_mask) - 1);
+      audio_channel_config_scheme =
+        "tag:dolby.com,2015:dash:audio_channel_configuration:2015";
```
[the remark this answers is not in the document]
Comment (ShaoWeiguo) [reply packager/mpd/base/xml/xml_node.cc:521]: If EC3 and AC4 can use MPEG scheme to represent the channel configuration, they use same scheme. 
If not, they use different scheme.

File: packager/mpd/base/media_info.proto (hunk 1 above)
Comment (ShaoWeiguo) [reply packager/mpd/base/media_info.proto:82]: Good suggestion. I modify locally and pass all the test cases. I will modify based on your suggestion.

File: packager/media/base/audio_stream_info.cc [hunk 5]
```diff
@@ -121,6 +123,14 @@ std::string AudioStreamInfo::GetCodecString(Codec codec,
       return "dts+";
     case kCodecEAC3:
       return "ec-3";
+    case kCodecAC4:
+      // ETSI TS 103 190-2 Digital Audio Compression (AC-4) Standard; Part 2:
+      // Immersive and personalized audio E.13. audio_object_type is composed of
+      // bitstream_version (3bits), presentation_version (2bits) and
+      // mdcompat (3bits)
```
Comment (kqyang) [inline packager/media/base/audio_stream_info.cc:130]: nit: add an ending period "."

File: packager/media/base/fourccs.h [hunk 6]
```diff
@@ -44,6 +44,7 @@ enum FourCC : uint32_t {
   FOURCC_dash = 0x64617368,
   FOURCC_ddts = 0x64647473,
   FOURCC_dec3 = 0x64656333,
+  FOURCC_dac4 = 0x64616334,
```
Comment (kqyang) [inline packager/media/base/fourccs.h:47]: The definition should be sorted. Can you move it after FOURCC_dac3?

File: packager/media/base/fourccs.h [hunk 7]
```diff
@@ -59,6 +60,7 @@ enum FourCC : uint32_t {
   FOURCC_dvvC = 0x64767643,
   FOURCC_ec_3 = 0x65632d33,  // "ec-3"
   FOURCC_ec3d = 0x65633364,
+  FOURCC_ac_4 = 0x61632d34,  // "ac-4"
```
Comment (kqyang) [inline packager/media/base/fourccs.h:63]: ditto. Can you move it after FOURCC_ac_3?

File: packager/media/base/stream_info.h [hunk 8]
```diff
@@ -49,6 +49,7 @@ enum Codec {
   kCodecDTSM,
   kCodecDTSP,
   kCodecEAC3,
+  kCodecAC4,
```
Comment (kqyang) [inline packager/media/base/stream_info.h:52]: The definition should be sorted. Can you move it after kCodecAC3?

File: packager/mpd/base/xml/xml_node.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (kqyang) [reply packager/mpd/base/xml/xml_node.cc:521]: Let me rephrase my question.

When the MPEG value is not available (== NO_MAPPING), the channel config scheme for EC-3 is `tag:dolby.com,2014:dash:audio_channel_configuration:2011`, but the channel config scheme for AC-4 is `tag:dolby.com,2015:dash:audio_channel_configuration:2015`.

Why are they different? 

I checked [DASH IF IOP v4.3](https://dashif.org/docs/DASH-IF-IOP-v4.3.pdf), it says `tag:dolby.com,2014:dash:audio_channel_configura31 tion:2011` should be used for both EC-3 and AC-4.

File: packager/media/base/audio_stream_info.cc (hunk 5 above)
Comment (ShaoWeiguo) [reply packager/media/base/audio_stream_info.cc:130]: Sure, thanks.

File: packager/media/base/fourccs.h (hunk 6 above)
Comment (ShaoWeiguo) [reply packager/media/base/fourccs.h:47]: Sure.

File: packager/media/base/fourccs.h (hunk 7 above)
Comment (ShaoWeiguo) [reply packager/media/base/fourccs.h:63]: Sure.

File: packager/media/base/stream_info.h (hunk 8 above)
Comment (ShaoWeiguo) [reply packager/media/base/stream_info.h:52]: Sure.

File: packager/mpd/base/xml/xml_node.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (ShaoWeiguo) [reply packager/mpd/base/xml/xml_node.cc:521]: Audio channel configuration for EC-3 and AC-4 in DASH IF IOP v4.3 is outdated. Dolby has changed it. Not only for internal guidance, but also try to fix in DASH-IF IOP. https://github.com/Dash-Industry-Forum/DASH-IF-IOP/issues/268

The document of **Audio Addendum and Corrigendum** for DASH IF IOP v4.3 is under review.  In this document, it states "tag:dolby.com,2014: dash:audio_channel_configuration:2011" is used for EC-3 and "tag:dolby.com,2015: dash:audio_channel_configuration:2015" is used for AC-4 when EC-3 and AC-4 can't use MPEG scheme to represent the channel configuration. 

From https://dashif.org/identifiers/audio_source_metadata/,  it says the @value attribute (tag:dolby.com,2014: dash:audio_channel_configuration:2011) shall contain a **four digit hexadecimal** representation of the 16 bit field that describes the channel assignment as defined by table E.5 in ETSI TS 102 366 [14]. 

From ETSI TS 103 190-2 V1.2.1 (2018-02) G.3.1, it says the @value attribute (tag:dolby.com,2015: dash:audio_channel_configuration:2015) shall be set to a **six-digit hexadecimal** representation of the 24-bit field
ac4_dsi_v1/ac4_presentation_v1_dsi/presentation_channel_mask_v1.

[the remark this answers is not in the document]
Comment (kqyang) [reply packager/mpd/base/xml/xml_node.cc:521]: I see. Thanks for the clarifications.

Frankly I think this design is a bit confusing. Anyway, do you mind add a clarification comment here, something like:

```c++
      // Note that the channel config schemes for EC-3 and AC-4 are different. 
      // See https://github.com/Dash-Industry-Forum/DASH-IF-IOP/issues/268.
      audio_channel_config_scheme =
        "tag:dolby.com,2015:dash:audio_channel_configuration:2015";
```

[the remark this answers is not in the document]
Comment (ShaoWeiguo) [reply packager/mpd/base/xml/xml_node.cc:521]: Sure. I will.

File: packager/hls/base/master_playlist.cc [hunk 9]
```diff
@@ -317,10 +317,17 @@ void BuildMediaTag(const MediaPlaylist& playlist,
     // for Apple Devices Appendices documents how to handle Dolby Digital Plus
     // JOC content.
     // https://developer.apple.com/documentation/http_live_streaming/hls_authoring_specification_for_apple_devices/hls_authoring_specification_for_apple_devices_appendices
+    // Dolby has qualified using IMSA to present AC4 immersive audio (IMS and
+    // CBI without object-based audio) for Dolby internal use only. IMSA is not
+    // included in any publicly-available specifications as of June, 2020.
```
[the remark this answers is not in the document]
Comment (ShaoWeiguo) [reply packager/hls/base/master_playlist.cc:322]: Sure. I will move the comments to make it easier to understand.

I expect the public specs can be available at the end of July.  For the player side, we plan to contribute to ExoPlayer firstly. And other players (like dash-js) are also in the scope.

**Comment by kqyang** [comment]
Thanks for addressing the comments. Can you update the new codec AC4 in the README page: https://github.com/google/shaka-packager/blob/master/README.md?

**Comment by ShaoWeiguo** [comment]
Sure. I will add new codec AC4 in the README page.

**Comment by kqyang** [comment]
Thanks @ShaoWeiguo for adding support for AC-4. The change has been merged.
